include/flexflow/memory_optimization.h

```diff
@@ -0,0 +1,166 @@
+/**
+ * @file memory_optimization.h
+ * @brief Memory optimization related stuff
+ *
+ * @copyright Copyright 2022 CMU, Facebook, LANL, MIT, and Stanford
+ * (alphabetical)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cassert>
+#include <string>
+
+namespace FlexFlow {
+
+enum class MemoryUsageType {
+  // Use global memory of a PCG as the measure of memory usage. No device
+  // mapping consideration.
+  GLOBAL,
+
+  // Use the max of peak per-device memory usage among devices as the measure.
+  // Need associated device mapping views.
+  PER_DEVICE_MAX,
+
+  // Use detailed per-device memory usage as the measure. Need associated device
+  // mapping views.
+  PER_DEVICE_ALL,
+};
+
+enum class MemorySearchAlgo {
+  // Multiple objective DP search. Combine memory cost and run time cost into
+  // one single cost function and add a factor to balance them.
+  MULTI_OBJECTIVE,
+};
+
+/**
+ * @brief Config class to control memory optimizations. This should be put into
+ * config.h and be stored in FFConfig. But for easy turnaround, put this here
+ * for now.
+ */
+class MemoryOptimConfig {
+public:
+  MemoryUsageType mem_usage_type;   ///< How to represent memory cost
+  MemorySearchAlgo mem_search_algo; ///< How to search for the optimal schedule
+  float run_time_cost_factor; ///< The weight factor of run time cost in the
+                              ///< overall cost function; used in
+                              ///< MULTI_OBJECTIVE algorithm
+
+  MemoryOptimConfig()
+      : mem_usage_type{MemoryUsageType::GLOBAL},
+        mem_search_algo{MemorySearchAlgo::MULTI_OBJECTIVE},
+        run_time_cost_factor{0.5} {}
+  MemoryOptimConfig(float factor)
+      : mem_usage_type{MemoryUsageType::GLOBAL},
+        mem_search_algo{MemorySearchAlgo::MULTI_OBJECTIVE},
+        run_time_cost_factor{factor} {}
+};
+
+/**
+ * @brief Hold the result (including memory information) of a graph_optimize on
+ * a PCG.
+ */
+class MemorySearchResult {
+public:
+  float run_time_cost{};
```

Comment:
Is the initialization here necessary?

Reply:
Yes, I think we need to init the float variables, especially in this codebase with C++11. We saw bugs due to that. Remember that we modified all parallel ops to init the sync_time, etc. which are all floats.